src/Rewards.sol

```diff
@@ -0,0 +1,181 @@
+     * @dev This constant defines the reward type.
+     * This should be kept consistent with the Reward struct.
+     */
+    bytes public constant REWARD_TYPE = "Reward(address recipient,uint256 amount,uint256 counter)";
+
+    /**
+     * @dev Reference to the NODL token contract.
+     */
+    NODL public nodlToken;
+
+    /**
+     * @dev Maximum amount of rewards that can be distributed in a period.
+     */
+    uint256 public rewardQuota;
+
+    /**
+     * @dev Duration of each reward period.
+     */
+    uint256 public rewardPeriod;
+
+    /**
+     * @dev Timestamp indicating when the reward quota is due to be renewed.
+     */
+    uint256 public quotaRenewalTimestamp;
+
+    /**
+     * @dev Amount of rewards claimed in the current period.
+     */
+    uint256 public rewardsClaimed;
+
+    /**
+     * @dev Address of the authorized oracle.
+     */
+    address public authorizedOracle;
+
+    /**
+     * @dev Mapping to store reward counters for each recipient to prevent replay attacks.
+     */
+    mapping(address => uint256) public rewardCounters;
+
+    /**
+     * @dev Struct on which basis an individual reward must be issued.
+     */
+    struct Reward {
+        address recipient;
+        uint256 amount;
+        uint256 counter;
+    }
+
+    /**
+     * @dev Error when the reward quota is exceeded.
+     */
+    error RewardQuotaExceeded();
+
+    /**
+     * @dev Error when the reward is not from the authorized oracle.
+     */
+    error UnauthorizedOracle();
+
+    /**
+     * @dev Error when the recipient's counter does not match.
+     */
+    error InvalidRecipientCounter();
+
+    /**
+     * @dev Event emitted when the reward quota is set.
+     */
+    event RewardQuotaSet(uint256 quota);
+
+    /**
+     * @dev Event emitted when a reward is minted.
+     */
+    event RewardMinted(address indexed recipient, uint256 amount, uint256 totalRewardsClaimed);
+
+    /**
+     * @dev Initializes the contract with the specified parameters.
+     * @param nodlTokenAddress Address of the NODL token contract.
+     * @param initialQuota Initial reward quota.
+     * @param initialPeriod Initial reward period.
+     * @param oracleAddress Address of the authorized oracle.
+     */
+    constructor(address nodlTokenAddress, uint256 initialQuota, uint256 initialPeriod, address oracleAddress)
+        EIP712(SIGNING_DOMAIN, SIGNATURE_VERSION)
+    {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        nodlToken = NODL(nodlTokenAddress);
+        rewardQuota = initialQuota;
+        rewardPeriod = initialPeriod;
+        rewardsClaimed = 0;
+        quotaRenewalTimestamp = block.timestamp + rewardPeriod;
+        authorizedOracle = oracleAddress;
+    }
+
+    /**
+     * @dev Sets the reward quota. Only accounts with the QUOTA_SETTER_ROLE can call this function.
+     * @param newQuota The new reward quota.
+     */
+    function setRewardQuota(uint256 newQuota) external {
+        _checkRole(QUOTA_SETTER_ROLE);
```

Comment:
You should use the admin role here, reading the contract it does not look like we need two different roles

Reply:
The admin can grant or revoke the quota setter role and then renounce its own role. This would be useful for the case the deployer is not the quota setter. Suppose the deployer is single easy to handle EOA but the quota setter is a multi-signature. You don't want to involve all those signers in the deployment process necessarily.

Reply:
The deployer can renounce renounce its role towards a multisig this wouldn't be an issue. Right now I don't quite see a use-case that would require us to keep those separate but let me know if you had anything concrete in mind.

Reply:
Alright, I agree